url73e3057d61_empeeoh_BACnet4J_tgz-pJ8-com_serotonin_bacnet4j_test_OverflowTestJ8/src/com/serotonin/bacnet4j/util/RequestUtils.java

```diff
@@ -0,0 +1,400 @@
+        for (final ObjectPropertyReference opr : oprs)
+            refs.add(opr.getObjectIdentifier(), opr.getPropertyIdentifier());
+
+        final PropertyValues pvs = readProperties(localDevice, rd, refs, callback);
+
+        // Read the properties in the same order.
+        final List<Pair<ObjectPropertyReference, Encodable>> results = new ArrayList<>();
+        for (final ObjectPropertyReference opr : oprs)
+            results.add(new ImmutablePair<ObjectPropertyReference, Encodable>(opr, pvs.getNoErrorCheck(opr)));
+
+        return results;
+    }
+
+    public static PropertyValues readProperties(final LocalDevice localDevice, 
+    											final RemoteDevice rd, 
+    											final PropertyReferences refs,
+    											final RequestListener callback) 
+    													throws BACnetException {
+        final PropertyValues propertyValues = new PropertyValues();
+        final RequestListenerUpdater updater = new RequestListenerUpdater(callback, propertyValues, refs.size());
+        boolean multipleSupported = rd.getServicesSupported() != null
+                && rd.getServicesSupported().isReadPropertyMultiple();
+
+        boolean forceMultiple = false;
+        // Check if a "special" property identifier is contained in the references.
+        for (List<PropertyReference> prs : refs.getProperties().values()) {
+            for (PropertyReference pr : prs) {
+                final PropertyIdentifier pi = pr.getPropertyIdentifier();
+                if (pi.equals(PropertyIdentifier.all) ||
+                	pi.equals(PropertyIdentifier.required) ||
+                    pi.equals(PropertyIdentifier.optional)) {
+                    forceMultiple = true;
+                    break;
+                }
+            }
+            if (forceMultiple)
+                break;
+        }
+
+        if (forceMultiple && !multipleSupported)
+            throw new BACnetException("Cannot send request to device " + rd.getInstanceNumber()
+            		 + "  ReadPropertyMultiple is required but not supported.");
+
+        if (forceMultiple || (refs.size() > 1 && multipleSupported)) {
+            // Read property multiple can be used. Determine the max references
+
+            int maxRef = rd.getMaxReadMultipleReferences();
+
+            // If the device supports read property multiple, send them all at once, or at least in partitions.
+            final List<PropertyReferences> partitions = refs.getPropertiesPartitioned(maxRef);
+            int counter = 0;
+            for (PropertyReferences partition : partitions) {
+            	final Map<ObjectIdentifier, List<PropertyReference>> properties = partition.getProperties();
+                final List<ReadAccessSpecification> specs = new ArrayList<>();
+                for (ObjectIdentifier oid : properties.keySet())
+                    specs.add(new ReadAccessSpecification(oid, new SequenceOf<PropertyReference>(properties.get(oid))));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.